docker/service.go

```diff
@@ -186,15 +188,48 @@ func (s *Service) up(imageName string, create bool) error {
 	}
 
 	return s.eachContainer(func(c *Container) error {
-		if outOfSync, err := c.OutOfSync(); err != nil {
-			return err
-		} else if outOfSync {
-			logrus.Warnf("%s needs rebuilding", s.Name())
+		if create {
+			if err := s.recreateIfNeeded(imageName, c); err != nil {
+				return err
+			}
 		}
+
 		return c.Up(imageName)
 	})
 }
 
+func (s *Service) recreateIfNeeded(imageName string, c *Container) error {
+	if s.context.NoRecreate {
+		return nil
+	}
+	outOfSync, err := c.OutOfSync(imageName)
+	if err != nil {
+		return err
+	}
+
+	containerInfo, err := c.findInfo()
+	if containerInfo == nil || err != nil {
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	logrus.WithFields(logrus.Fields{
+		"outOfSync":     outOfSync,
+		"ForceRecreate": s.context.ForceRecreate,
+		"NoRecreate":    s.context.NoRecreate}).Debug("Going to decide if recreate is needed")
+
+	if s.context.ForceRecreate || outOfSync {
+		logrus.Infof("Recreating %s", name)
+		if _, err := c.Recreate(imageName); err != nil {
+			return err
+		}
+	} else if outOfSync {
```

Comment:
Not sure I get this `if` condition : If `outOfSync` is true, no matter `ForceRecreate`, it got into the first one (and did the `Recreate`), and if `ForceRecreate` is true, it got into the first one too.
So I don't see how this `else if …` is ever gonna be executed. (I might miss something :sweat_smile:)